Refactor Oak Functions to separate encryption

tiziano88 opened the pull request:
Refactor the core logic of Oak Functions into a separate `OakFunctionsInstance` object. This object is still in the same crate as `OakFunctionsService` for now, but does not have any logic related to remote attestation nor encryption at all. Instead, that logic now only exists in the `OakFunctionsService` wrapper.

I changed the `InitializationState` from an enum to an `Option` so that it behaves like a proper functor (being able to use map and other methods on it).

`AttestationHandler` is changed to take an `FnOnce` and is expected to be constructed on the fly around a single invocation of an individual method, therefore it does not need to own the entire object being invoked, but can borrow a specific method when needed (and could be applied to several methods if necessary). The way around the ownership in the previous version was to have an `Arc<LookupDataManager>` instance shared between the outer service and the initialized instance, which is not a generally applicable pattern, while the current pattern should be fully generic.

The methods that do not require encryption defer to the inner methods directly (there may be a way of reducing the boilerplate in the future), while `invoke` creates a temporary `AttestationSessionHandler` object and uses it immediately for the invocation, then discards it.

Ref #4182

tiziano88 commented:
This is not ready for review, but I'd like some early feedback from @ipetr0v @conradgrobler @pmcgrath17 . Please read the description above first for context, or it won't make much sense. You can skip over the `OakFunctionsInstance` implementation for the most part, that's just the existing Oak Functions business logic.

conradgrobler reviewed (commented):
I think the approach is an improvement in general. I made a few comments related to preexisting naming, which I think became a bit more apparent with the refactoring.

ipetr0v commented:
> This object is still in the same crate as `OakFunctionsService` for now

I think we should split crates only when it's strongly necessary (i.e. the crate is reused by multiple crates). So that we won't have too many crates in the repo.

tiziano88 commented:
> > This object is still in the same crate as `OakFunctionsService` for now
> 
> I think we should split crates only when it's strongly necessary (i.e. the crate is reused by multiple crates). So that we won't have too many crates in the repo.

Right, this is not split yet, but will be split in the future, because it will be used from two different Oak Functions binaries (each with its own crate): one for Oak Restricted Kernel, one for Linux.

Anyways, that's not the main point of me asking for the review, I would like more feedback around the encryption logic if possible, thanks.

ipetr0v commented:
I think the abstraction of having a trusted application logic that creates a handler that is called in the attestation handler may be too convoluted for devs to use, for example they would need to fight the Rust borrow checker by creating a handler that can access everything (i.e. handler lifetimes etc).

I thinkg this abstraction could be swapped. We can have a `TrustedApplicationEngine` that will receive an encrypted request, decrypt it and call the application logic. Though this might be complicated, since we have a gRPC interface specific to each app.

So even simpler solution will be to make devs just use the `Encryptor`/`Decryptor` objects to encrypt and decrypt the data.

tiziano88 commented:
> So even simpler solution will be to make devs just use the `Encryptor`/`Decryptor` objects to encrypt and decrypt the data.

I agree, we should have a trait with the method to create those object from a public key as part of the first message in the stream; we can then implement this trait for Oak Restricted Kernel and Oak Containers separately, and inject the appropriate concrete object in the implementation of the business logic in order to get two versions of the binary. I think that can be a good next step after this PR (hopefully this makes that a bit simpler).

pmcgrath17 commented on oak_functions_service/src/instance.rs:
```diff
@@ -0,0 +1,111 @@
+    type Error = anyhow::Error;
+
+    fn invoke(&mut self, request_bytes: &[u8]) -> Result<Vec<u8>, Self::Error> {
+        self.invoke(request_bytes)
```
+1 on using the handle_user_request name

pmcgrath17 reviewed (approved):
I'm not as familiar with Rust, but at a high level this change does make sense and appears to be an improvement.

One general question (from someone who is still fairly new to the Oak codebase): can more comments be added above functions indicating what is being done? From the names of the functions alone, it isn't always apparent to me.

tiziano88 commented:
> I'm not as familiar with Rust, but at a high level this change does make sense and appears to be an improvement.
> 
> One general question (from someone who is still fairly new to the Oak codebase): can more comments be added above functions indicating what is being done? From the names of the functions alone, it isn't always apparent to me.

I removed some of the unnecessary functions (specifically a couple of the `invoke` ones that are now not needed).

Also added some doc links, but for them to be useful, #4266 needs to be fixed first.

I'll see if there something more I can add in the meanwhile.
